setup.go

```diff
@@ -155,19 +149,103 @@ func setup(c *caddy.Controller) error {
 				return c.ArgErr()
 			}
 			if timeout < 0 {
-				return errors.New("Parse error: dial_timeout cannot be negative.")
+				return c.Err("dial_timeout cannot be negative.")
 			}
 			fp.dialTimeout = time.Second * time.Duration(timeout)
 		case "upstream":
 			if len(args) != 1 {
 				return c.ArgErr()
 			}
 			fp.upstream = args[0]
+		case "acl":
+			if len(args) != 0 {
+				return c.Err("acl should be only subdirective on the line")
+			}
+			args := c.RemainingArgs()
+			if len(args) > 0 {
+				return c.ArgErr()
+			}
+			c.Next()
+			if c.Val() != "{" {
+				return c.Err("acl directive must be followed by opening curly braces \"{\"")
+			}
+			for {
+				if !c.Next() {
+					return c.Err("acl blockmust be ended by closing curly braces \"}\"")
+				}
+				aclDirective := c.Val()
+				args := c.RemainingArgs()
+				if aclDirective == "}" {
+					break
+				}
+				if len(args) == 0 {
+					return c.ArgErr()
+				}
+				var ruleSubjects []string
+				var err error
+				aclAllow := false
+				switch aclDirective {
+				case "allow":
+					ruleSubjects = args[:]
+					aclAllow = true
+				case "allowfile":
+					if len(args) != 1 {
+						return c.Err("allowfile accepts a single filename argument")
+					}
+					ruleSubjects, err = readLinesFromFile(args[0])
+					if err != nil {
+						return err
+					}
+					aclAllow = true
+				case "deny":
+					ruleSubjects = args[:]
+				case "denyfile":
+					if len(args) != 1 {
+						return c.Err("denyfile accepts a single filename argument")
+					}
+					ruleSubjects, err = readLinesFromFile(args[0])
+					if err != nil {
+						return err
+					}
+				default:
+					return c.Err("expected acl directive: allow/allowfile/deny/denyfile." +
+						"got: " + aclDirective)
+				}
+				for _, rs := range ruleSubjects {
+					ar, err := newAclRule(rs, aclAllow)
+					if err != nil {
+						return err
+					}
+					fp.aclRules = append(fp.aclRules, ar)
+				}
+			}
 		default:
 			return c.ArgErr()
 		}
 	}
 
+	if fp.upstream != "" && (fp.aclRules != nil || len(fp.whitelistedPorts) != 0) {
+		return c.Err("upstream subdirective is incompatible with acl/ports subdirectives")
+	}
+
+	if fp.aclRules == nil {
```

Comment:
I don't have a strong opinion, but that might be a bit unexpected, and it might be a good idea to shoot for convincing the user to build upon default rules (which we provide in README)

Reply:
I think the clearest behavior is to have the default rules automatically, unconditionally appended to the user's rules.  Otherwise, things get confusing.  In the current behavior
 * empty ACL is not equivalent to no ACL
 * Adding `acl { deny 8.8.8.8 }` to your config doesn't just disable access to 8.8.8.8 ... it also enables access to all localhost/LAN IPs.

If the user wants to override the default rules, they can end their rules with `allow all` or `deny all`, like you're already recommending in the documentation.  That way, our safeguards are in place unless explicitly disabled.

Reply:
I thought we're making both of those illegal, since they lack `deny all`/`allow all` rules.
And if user writes
```
acl {
  deny 8.8.8.8
  allow all
}
```
that makes it less unexpected that everything else is allowed now.

Reply:
I agree that requiring an "all" rule is nice and explicit.  However, it encourages users to write simple rules like the one you've included, which is clear ... but "unsafe".  Writing a "safe" ACL now requires copy-pasting the default rules out of the documentation, which seems unfortunate to me.

If you want the rules to be very explicit, one option would be to add a fifth rule type: "default".  Then every ACL should end with "allow all", "deny all", or "default".